lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Corres_Cases
+imports Corres_UL
+begin
+
+text \<open>
+  This file defines the following main methods for safe data type case distinctions on
+  corres/corres_underlying predicates.
+
+  \<^item> corres_cases_left: case distinction on abstract monad
+  \<^item> corres_cases_right: case distinction on concrete monad
+  \<^item> corres_cases: case distinction first of abstract/concrete monad
+  \<^item> corres_cases_both: simultaneous (quadratic) case distinction on both sides, with safe
+                       elimination of trivially contradictory cases.
+
+  The first 3 methods take no arguments, corres_cases_both takes an optional simp argument to,
+  for example, unfold relations that synchronises cases between abstract and concrete side.
+
+  The case distinctions work if the entire monad is a "case" statement, or if the monad is a
+  @{const bind} or @{const bindE} term with a "case" statement in the head position.
+
+  There is an existing method for case distinctions (@{method wpc}), but this method is not
+  flexible enough for @{term corres}. The reason for this is that @{method wpc}
+  will always treat abstract and concrete guards the same way. This doesn't always work, because
+  the abstract guard usually only depends on abstract values and the concrete guard only on
+  concrete values. If we make a case distinction on such an abstract value, the concrete guard
+  cannot to transformed into one that follows that case distinction, so @{method wpc} will fail.
```

Comment:
"cannot be transformed"?

Reply:
bit hazy on meaning of "the concrete guard cannot to transformed into one that follows that case distinction"

Reply:
I was trying to be concise, but that's probably a mistake. The situation is that for terms like

```
  "⋀x. corres .. (?G x) ?G' (case x of ..) m"
```

if you do case distinction on `x`, you can transform `?G` into e.g. `λx s. (x = None ⟶ ?G1 x s) ∧ (∀y. x = Some y ⟶ ?G2 x y s)`, but you can't do that with `?G'` because `?G'` doesn't depend on `x`. The best you can get for `?G'` is just `?G1' and ?G2'`.

Reply:
Ah, that makes sense. I think the original sentence is a bit over-compressed, explaining what "following" "that" case distinction means definitely helped me here.